### ISSUE TITLE
Up to date tips on using AppSync Offline in production

### ISSUE THREAD

**Issue by PeteDuncanson** [aws-amplify/amplify-js#9481]
I've been advised to use AppSync Offline rather than DataStore due to scaling issues we where hitting with DS.

We've spend quite some time swapping it all out in our project but now we have hit a new issue (again with scale).

It seems that AppSync Offline out of the box uses an in memory storage to act as the cache. We are using the generated GraphQL queries for speed of development and our models have some connections which leads to some nesting of models. As a result our queries and responses are quite large in size. After some clicking round we are quickly filling up the storage and getting errors about it being out of space, see https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/168.

The google found fix for this (https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/168#issuecomment-448830161) is to install something called localforage which we did and indeed the space issue went away! However after some more testing what we found is localforage doesn't seem to do any expiring of the data it caches so it just keeps on growing. If this was using the IndexedDB key/store as I assumed it should this wouldn't be an issue but instead it seems to just be stringifying a forever growing JSON object everytime its touched which quickly brings the browser to a crawl and eventual crash.

I can't find any settings in localforages config to enable a way to the amount of data we store or how to expire it. There is a size limit but thats for WebSQL and that is no longer recommended to use now IndexedDB is available in all browsers.

The only hint of a solution to this is a bit of a hack which is to just "clear the cache at the start of the day" which could work...but I would need to check first to see if we have any saved mutations queued up which need to go out first before we flush the cache. Doable but more work and it just feels like this is a problem that should have been solved before.

https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/125#issuecomment-397137476
 
Given that I've followed all the advice and recommendations that I can officially find I'm now once more stuck and in a bit of a dead end and can't use AppSync Offline currently.

Any pointers or can anyone work with me on how to find a new best practise for this? I can't be the only to have found or hit this issue in the last 3-4 years surely?

**Amplify CLI Version** 

7.6.8

**Provide additional details e.g. code snippets**

Here is how I'm configuring AppSync in my client side code:

```
const client = new AWSAppSyncClient(
  {
    url: AppSyncConfig.aws_appsync_graphqlEndpoint,
    region: AppSyncConfig.aws_appsync_region,
    auth: {
      type: AUTH_TYPE[AppSyncConfig.aws_appsync_authenticationType],
      apiKey: AppSyncConfig.aws_appsync_apiKey,
      jwtToken: async () =>
        (await Auth.currentSession()).getAccessToken().getJwtToken(), 
    
    },
    disableOffline: false, // Set this to true if you want to avoid the local storage hogging error, false if you want to start caching stuff for offline use
    offlineConfig: {
      storage: localForage,
    },
  },
  {
    defaultOptions: {
      watchQuery: {
        fetchPolicy: "cache-and-network" 
        errorPolicy: "ignore",
      },
      query: {
        fetchPolicy: "network-only",
        errorPolicy: "all",
      },
    },
  }
);
```

**Comment by nubpro** [comment]
Pete is offline persistent a requirement for your project?

Because I honestly dont see the its worth the headache to fumble around DataStore or any other similar offline strategies at their current state. On top of that, AppSync SDK is seemingly out of maintenance. You will be own your own.

**Comment by PeteDuncanson** [comment]
@nubpro yep, I 100% need it.

We've dumped DataStore completely now (I'll never get back all those weeks of dev time sadly) and AppSync Offline was the replacement recommended to me. However like you state its seems to have not been touched in years :(

**Comment by PeteDuncanson** [comment]
Been poking around today and been amazed at the size of some of the auto generated queries, one of our mutations has over 1000 lines of fields to return! So easy for these to go unnoticed but I can easily see why we might be hitting size issues with this much data flying around.

Pondering hand rolling/tuning our queries to reduce that pay load. I'm finding all this out for the first time though so if nothing else I'd like this issue to be a catch all for how to deal with Offline well with Amplify AppSync if possible so others coming behind don't have to go through the same pain.

In short would still love some input on this as I think it deserves some fresh eyes as DataStore isn't the magic bullet for all use cases.

**Comment by PeteDuncanson** [comment]
I had a DM within Discord about if I had an update on this one so thought I would do that here so others could get it.

Reviewing the limited number of models that we actually needed to store/update I chose to roll my own caching layer. This is far from what I wanted to do but seemed like the quickest solution out of the hole I was in. Its not generic and requires some hardcorded knowledge but it works and has been working since we rolled it out in Feb this year.

A quick crash course if anyone wants to follow in my footsteps.

The biggest issue seemed to be the generic way that all the storage solutions tried to work, save down a hugh JSON object as they couldn't easily know how to break your data up. I went a different way as I know how my data related so could create a table per model and manage stitching it all together myself. As a result we never hit any size issues or cooked the CPU fans trying to stringify huge JSON objects. 

Here is roughly how I did it:

* Used idb (https://github.com/jakearchibald/idb) to ease interacting with the IndexedDB locally (this gives you promises for IndexedDB which was so much easier).
* Created a "cache" database with a table per model that we wanted to store locally
* Created a "mutation" database with a table per model that we wanted to be push up
* Then we had a Service layer that everything called with managed all our CRUD as far as the UI was concerned. Under the hood this called out to are CachingLayer (which just updated/added the models into the right table) or our MutationLayer which would either do the update there and then if we were online or add it to the queue of updates to go up next time we are online. 
* Mutations are just the payload of the mutation that we would send to GraphQL stored in the correct table we created earlier.
* When ever we do a read, if we are online we just go to the network and do a GraphQL query as normal and then cache the response if its good. If offline then we just go to the cache.
* When our online state changes to "online" we check if there are any mutations in our mutation queue, if so we process them and fire them up to the GraphQL endpoint. If the order of the updates is important (such as a nested object) then we process the mutations in a set order, in our case starting with leaf nodes and working our way back up the nesting until all done.
* We de-nest any nested responses, so if we get back an Visit which has a list of Jobs then we cache each Job keyed by its ID and then added in a new field to the Visit with an array of JobIDs. Then when we request these from cache we rebuild the list of Jobs using the IDs to look them up in the cache. This seems to work well and doesn't have any issues performance wise.
* We have the option to clear the cache at the start of the day.
* We have a button that will clear the cache manually within our site (handy if anything quirky does go wrong with a release or if our schema changes) 
* We have the option to force the site to run in "offline mode" if the connection is flakey or limited, makes it all run a lot quicker.

So far this has worked for us but its a big investment, took me a good few days to get this up and running.

Ideally I'd like to try to make a more generic version of this but before you know it you are recreating one of the caching/query frameworks that are already out there. I just wanted something lightweight that I knew would do the job and not have yet more surprises instore for me!

Going forward hand rolling your GraphQL queries is the next big win, that greatly reduces the payload size.

Hope that helps someone for now!

Pete

**Comment by chrisbonifacio** [comment]
Closing this issue as the `aws-appsync` is no longer the recommended approach to fulfill offline use cases in the Amplify platform.

From the [aws-mobile-appsync-sdk-js](https://github.com/awslabs/aws-mobile-appsync-sdk-js) readme

> For front-end web and mobile development, we recommend using the [Amplify](https://aws-amplify.github.io/) clients which are optimized to connect to the AppSync backend.
> 
> For DynamoDB data sources, use the DataStore category in the Amplify client. It provides the best developer experience and built-in conflict detection and resolution.
> For non-DynamoDB data sources in scenarios where you have no offline requirements, use the API (GraphQL) category in the Amplify client.

**Comment by PeteDuncanson** [comment]
Just to add to this, DataStore isn't a great off line solution at all, its just the only one AWS Amplify has now so everyone gets pointed to that but it is a massive foot cannon awaiting to burn weeks of your develment time at some future point. 

So if you need to do anything that is beyond DataStore (which most Apps do in my experience) then you have to go the GraphQL route which has no offline support out of the box. So you are going to have to roll your own. Joy.
